### PR TITLE
Service Mesh mTLS: introduce auth map

### PR DESCRIPTION
**Pull request by mhofstetter**
Currently, authentication state (if required by a policy) is tracked as property in the conntrack map which requires authentication in the userspace on a per connection basis.

These changes introduce the bpf map `auth` which enables keeping track of authentication state between security identities in combination with the remote node id, auth type and an expiration.

### CODE REVIEW

File: bpf/lib/common.h [hunk 1]
```diff
@@ -341,6 +341,18 @@ struct policy_entry {
...
+struct auth_key {
+	__u32       local_sec_label;
+	__u32       remote_sec_label;
+	__u16       remote_host_id;
+	__u8        auth_type;
+	__u8        pad;
+};
+
+struct auth_info {
+	__s64       expiration;
```
Comment (jrajahalme) [inline bpf/lib/common.h]: This needs to be unsigned rather than signed type (i.e., `__u64`).

File: cilium/cmd/bpf_auth_list.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	"github.com/cilium/cilium/pkg/maps/auth"
+	"github.com/cilium/cilium/pkg/policy"
+)
+
+type authEntry struct {
+	LocalIdentity  uint32
+	RemoteIdentity uint32
+	RemoteNodeID   uint16
+	AuthType       uint8
+	Expiration     int64
```
Comment (jrajahalme) [inline cilium/cmd/bpf_auth_list.go]: We'll want to format this as human readable time at some point, but this is fine for now.

File: pkg/maps/auth/auth_map.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	AuthType       uint8  `align:"auth_type"`
+	Pad            uint8  `align:"pad"`
+}
+
+func (r *AuthKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(r) }
+
+func (r *AuthKey) NewValue() bpf.MapValue { return &AuthInfo{} }
+
+func (r *AuthKey) String() string {
+	return fmt.Sprintf("%d --> %d (%d)", r.LocalIdentity, r.RemoteIdentity, r.RemoteNodeID)
```
Comment (jrajahalme) [inline pkg/maps/auth/auth_map.go]: This is mutual auth, so maybe `<->` instead of `-->` ?

File: pkg/maps/auth/auth_map.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		RemoteIdentity: remoteIdentity.Uint32(),
+		RemoteNodeID:   remoteNodeID,
+		AuthType:       authType.Uint8(),
+	}
+}
+
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapValue
+type AuthInfo struct {
+	Expiration int64 `align:"expiration"`
```
Comment (jrajahalme) [inline pkg/maps/auth/auth_map.go]: `uint64` here as well.

**Review by jrajahalme — APPROVED** [review]
Small nits only :-)

File: pkg/maps/auth/auth_map.go (hunk 3 above)
Comment (mhofstetter) [reply pkg/maps/auth/auth_map.go]: changed it to a uniform property based version -> `localIdentity=..., remoteIdentity=...`

File: cilium/cmd/bpf_auth_list.go (hunk 2 above)
Comment (mhofstetter) [reply cilium/cmd/bpf_auth_list.go]: agree. for the time being it's really more to inspect the content of the map.

**Comment by jrajahalme** [comment]
New map types need to be added to the alignchecker like for example in commit 16e132ee4fb60a382c7bb2504a5f697317c270cf, both in `bpf/bpf_alignchecker.c` and `pkg/datapath/alignchecker/alignchecker.go`

**Review by dylandreimerink — APPROVED** [review]
LGTM!

**Comment by mhofstetter** [comment]
E2E Tests failed with 

```
Found 2 k8s-app=cilium logs matching list of errors that must be investigated:
[2023-03-09T17:00:37.178Z]     2023-03-09T17:00:16.900966073Z level=warning msg="Skipping symbol substitution" subsys=elf symbol=cilium_auth_map
...
```

added `cilium_auth_map` to `ignoredELFPrefixes`

File: Documentation/spelling_wordlist.txt [hunk 5]
```diff
@@ -187,6 +187,7 @@ artii
 asm
 assignees
 au
+auth
```
Comment (qmonnet) [inline Documentation/spelling_wordlist.txt:190]: Can you please fold this change with the commit that introduces the word in the documentation?

File: pkg/maps/auth/auth_map.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/ebpf"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/policy"
+)
+
+const (
+	MapName    = "cilium_auth_map"
+	MaxEntries = 512000
```
Comment (qmonnet) [inline pkg/maps/auth/auth_map.go]: I think the max size for most eBPF maps is configurable using dedicated options for the agent. Would it make sense to have the same here?

Please update `Documentation/network/ebpf/maps.rst`.

**Review by qmonnet — CHANGES_REQUESTED** [review]

File: Documentation/spelling_wordlist.txt (hunk 5 above)
Comment (mhofstetter) [reply Documentation/spelling_wordlist.txt:190]: of course, done

File: pkg/maps/auth/auth_map.go (hunk 6 above)
Comment (mhofstetter) [reply pkg/maps/auth/auth_map.go]: @qmonnet thanks for the valuable input.

the max size of the auth map is now configurable and documented!

**Comment by mhofstetter** [comment]
rebased to master and resolved conflicts in generated documentation files (without any other changes)

**Review by qmonnet — APPROVED** [review]
Looks good to me, thank you!

**Review by meyskens — APPROVED** [review]
LGTM for me

**Pull request merged**